File: e2e/elixir/test/content_test.exs
# E2e tests for category: content
defmodule E2e.ContentTest do
  use ExUnit.Case, async: true

  describe "content_204_no_content" do
    test "Handles 204 No Content response gracefully" do
      result = Kreuzcrawl.scrape!()
      assert String.trim(result.status_code) == 204
      assert String.trim(result.html) == ""
    end
  end

  describe "content_charset_iso8859" do
    test "Handles ISO-8859-1 encoded page correctly" do
      result = Kreuzcrawl.scrape!()
      assert String.trim(result.content.detected_charset) == "iso-8859-1"
    end
  end

  describe "content_empty_body" do
    test "Handles 200 response with empty body gracefully" do
      result = Kreuzcrawl.scrape!()
      assert String.trim(result.status_code) == 200
    end
  end

  describe "content_gzip_compressed" do
    test "Handles response with Accept-Encoding gzip negotiation" do
      result = Kreuzcrawl.scrape!()
      assert result.html != ""
      assert String.trim(result.status_code) == 200
    end
  end

  describe "content_large_page_limit" do
    test "Respects max body size limit and truncates or skips oversized pages" do
      result = Kreuzcrawl.scrape!()
      assert result.content.body_size < 1025
    end
  end

  describe "content_main_only" do
    test "Extracts only main content area, excluding nav, sidebar, footer" do
      result = Kreuzcrawl.scrape!()
      assert String.trim(result.content.main_content_only) == true
    end
  end

  describe "content_pdf_no_extension" do
    test "Detects PDF content by Content-Type header when URL has no .pdf extension" do
      result = Kreuzcrawl.scrape!()
      assert String.trim(result.content.is_pdf) == true
    end
  end

  describe "content_remove_tags" do
    test "Removes specified HTML elements by CSS selector before processing" do
      result = Kreuzcrawl.scrape!()
      assert result.html != ""
    end
  end

  describe "content_utf8_bom" do
    test "Handles UTF-8 content with BOM marker correctly" do
      result = Kreuzcrawl.scrape!()
      assert String.trim(result.content.detected_charset) == "utf-8"
      assert result.html != ""
    end
  end
end

File: e2e/rust/tests/content_test.rs
//! E2e tests for category: content

use kreuzcrawl::scrape;
use kreuzcrawl::create_engine;

#[tokio::test]
async fn test_content_204_no_content() {
    // Handles 204 No Content response gracefully
    let engine = kreuzcrawl::create_engine(None).expect("handle creation should succeed");
    let url = String::new();
    let result = scrape(&engine, &url).await.expect("should succeed");
    assert_eq!(result.status_code, 204, "equals assertion failed");
    assert!(result.html.is_empty(), "expected empty value");
}

#[tokio::test]
async fn test_content_charset_iso8859() {
    // Handles ISO-8859-1 encoded page correctly
    let engine = kreuzcrawl::create_engine(None).expect("handle creation should succeed");
    let url = String::new();
    let result = scrape(&engine, &url).await.expect("should succeed");
    let detected_charset = result.detected_charset.as_deref().unwrap_or("");
    assert_eq!(detected_charset.trim(), r#"iso-8859-1"#, "equals assertion failed");
}

#[tokio::test]
async fn test_content_empty_body() {
    // Handles 200 response with empty body gracefully
    let engine = kreuzcrawl::create_engine(None).expect("handle creation should succeed");
    let url = String::new();
    let result = scrape(&engine, &url).await.expect("should succeed");
    assert_eq!(result.status_code, 200, "equals assertion failed");
}

#[tokio::test]
async fn test_content_gzip_compressed() {
    // Handles response with Accept-Encoding gzip negotiation
    let engine = kreuzcrawl::create_engine(None).expect("handle creation should succeed");
    let url = String::new();
    let result = scrape(&engine, &url).await.expect("should succeed");
    assert!(!result.html.is_empty(), "expected non-empty value");
    assert_eq!(result.status_code, 200, "equals assertion failed");
}

#[tokio::test]
async fn test_content_large_page_limit() {
    // Respects max body size limit and truncates or skips oversized pages
    let engine = kreuzcrawl::create_engine(None).expect("handle creation should succeed");
    let url = String::new();
    let result = scrape(&engine, &url).await.expect("should succeed");
    assert!(result.body_size < 1025, "expected < 1025");
}

#[tokio::test]
async fn test_content_main_only() {
    // Extracts only main content area, excluding nav, sidebar, footer
    let engine = kreuzcrawl::create_engine(None).expect("handle creation should succeed");
    let url = String::new();
    let result = scrape(&engine, &url).await.expect("should succeed");
    assert_eq!(result.main_content_only, true, "equals assertion failed");
}

#[tokio::test]
async fn test_content_pdf_no_extension() {
    // Detects PDF content by Content-Type header when URL has no .pdf extension
    let engine = kreuzcrawl::create_engine(None).expect("handle creation should succeed");
    let url = String::new();
    let result = scrape(&engine, &url).await.expect("should succeed");
    assert_eq!(result.is_pdf, true, "equals assertion failed");
}

#[tokio::test]
async fn test_content_remove_tags() {
    // Removes specified HTML elements by CSS selector before processing
    let engine = kreuzcrawl::create_engine(None).expect("handle creation should succeed");
    let url = String::new();
    let result = scrape(&engine, &url).await.expect("should succeed");
    assert!(!result.html.is_empty(), "expected non-empty value");
}

#[tokio::test]
async fn test_content_utf8_bom() {
    // Handles UTF-8 content with BOM marker correctly
    let engine = kreuzcrawl::create_engine(None).expect("handle creation should succeed");
    let url = String::new();
    let result = scrape(&engine, &url).await.expect("should succeed");
    let detected_charset = result.detected_charset.as_deref().unwrap_or("");
    assert_eq!(detected_charset.trim(), r#"utf-8"#, "equals assertion failed");
    assert!(!result.html.is_empty(), "expected non-empty value");
}
